OpenMage/magento-lts
# Add translation helper shell script

### Description (*)

Because in #2317 I lost track of which translation strings I introduced, I wrote a quick helper script for me to find strings used in the `__` function or in XML files but not defined in a CSV file. I thought it would be useful so I fixed it up a bit and added it as `shell/translations.php`.

It has two modes, `missing` which will tell you which strings you must add to a CSV file, and `unused` which will print strings defined in CSV files but not apparently used anywhere in the source. Additionally, it lets you pipe in a list of files to only check, which is useful to test PRs.

```
Usage:  php -f translations.php -- [options]
        php -f translations.php -- missing --verbose

  missing           Display used translations strings that are missing from csv files
  unused            Display defined translations strings that are not used in templates
  --verbose         Include filename with output
  --lang <lang>     Specify which language pack to check in app/locale, default is en_US
  deprecated        Find deprecated usage of the global __() function
  --fix             Overwrite files to fix deprecated usage DO NOT RUN IN PRODUCTION!
  help              This help

Note: By default, this script will check all files in this repository. However,
      you can pipe a list of files to check for missing translations strings.
      This is useful for checking a specific commit. For example:

      # Check if last two commits may have introduced missing translations
      git diff --name-only HEAD~2 | php -f translations.php -- missing
```

### Related Pull Requests


### Fixed Issues (if relevant)


### Manual testing scenarios (*)

1. Run `php shell/translations.php missing --verbose` to see php, phtml, or xml files that used non-defined translations
1. Run `php shell/translations.php unused --verbose` to see CSV files that have unused translations

### Questions or comments
```
# Returns 422 unused strings
php shell/translations.php unused | wc -l

# Returns 85 missing strings
php shell/translations.php missing | wc -l
```

It could be possible that the code to find usage of the `__` function is not perfect, such as `__('Total: ' . $total)` would only pick up `'Total: '`, but I don't think that's used often, or shouldn't be, as the `%s` usage is preferred.


### Contribution checklist (*)
